src/Grpc.StatusProto/RpcStatusExtensions.cs

```diff
@@ -0,0 +1,263 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Core;
+
+namespace Grpc.StatusProto;
+
+/// <summary>
+/// Extensions for <see cref="Google.Rpc.Status"/> to retrieve detailed error information.
+/// Based on ideas from:
+/// https://github.com/googleapis/gax-dotnet/blob/main/Google.Api.Gax.Grpc/RpcExceptionExtensions.cs
+/// </summary>
+public static class RpcStatusExtensions
+{
+    /// <summary>
+    /// Cache the full names of the messages types
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    private static class MessageNameCache<T> where T : class, IMessage<T>, new()
+    {
+        public static readonly string FullName = new T().Descriptor.FullName;
+    }
+
+    /// <summary>
+    /// Retrieves the error details of type <typeparamref name="T"/> from the <see cref="Google.Rpc.Status"/>
+    /// message.
+    /// Note: experimental API that can change or be removed without any prior notice.
+    /// </summary>
+    /// <remarks>
+    /// <example>
+    /// For example, to retrieve any <see cref="Google.Rpc.ErrorInfo"/> that might be in the status details:
+    /// <code>
+    ///   var errorInfo = status.GetStatusDetail&lt;ErrorInfo&gt;();
+    ///   if (errorInfo is not null) {
+    ///      // ...
+    ///   }
+    /// </code>
+    /// </example>
+    /// </remarks>
+    /// <typeparam name="T">The message type to decode from within the error details.</typeparam>
+    /// <param name="status">The RPC status to retrieve details from. Must not be null.</param>
+    /// <returns>The first error details of type <typeparamref name="T"/> found, or null if not present</returns>
+    public static T? GetStatusDetail<T>(this Google.Rpc.Status status) where T : class, IMessage<T>, new()
+    {
+        //TODO(tonydnewell) maybe move this method to Google.Api.CommonProtos
```

Comment:
If we move it, the namespace would be Google.Api.

The conflict is definitely a potential cause of annoyance - but it'll be *reasonably* temporary as I'll remove it from GAX in our next major version. If we can think of an equally-good name, I'd be fine to have the two different names... but if `GetStatusDetail` is really the definitely-best name, I'd rather take the short-term ambiguity hit (and probably write a paragraph about it in our client library help) than lose the best name.

Reply:
Ah - if it's on Google.Rpc.Status, rather than the current RpcException, then just `GetDetail` might be better - it's more concise and it avoids the conflict. Thoughts?

Reply:
I'll change to GetDetail as it is already on the Status.  I've not changed the namespace unless you think we should start using the Google.Rpc namespace now before any methods are moved?  Also I'll not be too upset if the methods are never moved.

Reply:
I guess if you change the namespace to reflect where the methods will be, then all the calling code will already be "right". And I do want to put the extension methods in Google.Api.CommonProtos. Happy to create a PR for that myself, or you can.